x/stake/handler.go

```diff
@@ -0,0 +1,318 @@
+	k.setCandidate(ctx, candidate)
+	return nil
+}
+
+// Perform all the actions required to bond tokens to a delegator bond from their account
+func (k Keeper) BondCoins(ctx sdk.Context, bond DelegatorBond, candidate Candidate, amount sdk.Coin) sdk.Error {
+
+	_, err := k.coinKeeper.SubtractCoins(ctx, bond.DelegatorAddr, sdk.Coins{amount})
+	if err != nil {
+		return err
+	}
+	newShares := k.candidateAddTokens(ctx, candidate, amount.Amount)
+	bond.Shares = bond.Shares.Add(newShares)
+	k.setDelegatorBond(ctx, bond)
+	return nil
+}
+
+func (k Keeper) handleMsgUnbond(ctx sdk.Context, msg MsgUnbond) sdk.Result {
+
+	// check if bond has any shares in it unbond
+	bond, found := k.getDelegatorBond(ctx, msg.DelegatorAddr, msg.CandidateAddr)
+	if !found {
+		return ErrNoDelegatorForAddress().Result()
+	}
+	if !bond.Shares.GT(sdk.ZeroRat) { // bond shares < msg shares
+		return ErrInsufficientFunds().Result()
+	}
+
+	// if shares set to special case Max then we're good
+	if msg.Shares != "MAX" {
+		// test getting rational number from decimal provided
+		shares, err := sdk.NewRatFromDecimal(msg.Shares)
+		if err != nil {
+			return err.Result()
+		}
+
+		// test that there are enough shares to unbond
+		if !bond.Shares.GT(shares) {
+			return ErrNotEnoughBondShares(msg.Shares).Result()
+		}
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasUnbond,
+		}
+	}
+
+	// retrieve the amount of bonds to remove (TODO remove redundancy already serialized)
+	var shares sdk.Rat
+	var err sdk.Error
+	if msg.Shares == "MAX" {
+		shares = bond.Shares
+	} else {
+		shares, err = sdk.NewRatFromDecimal(msg.Shares)
+		if err != nil {
+			return err.Result()
+		}
+	}
+
+	// subtract bond tokens from delegator bond
+	if bond.Shares.LT(shares) { // bond shares < msg shares
+		return ErrInsufficientFunds().Result()
+	}
+	bond.Shares = bond.Shares.Sub(shares)
+
+	// get pubKey candidate
+	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if !found {
+		return ErrNoCandidateForAddress().Result()
+	}
+
+	revokeCandidacy := false
+	if bond.Shares.IsZero() {
+
+		// if the bond is the owner of the candidate then
+		// trigger a revoke candidacy
+		if bytes.Equal(bond.DelegatorAddr, candidate.Address) &&
+			candidate.Status != Revoked {
+			revokeCandidacy = true
+		}
+
+		// remove the bond
+		k.removeDelegatorBond(ctx, bond)
+	} else {
+		k.setDelegatorBond(ctx, bond)
+	}
+
+	// Add the coins
+	returnAmount := k.candidateRemoveShares(ctx, candidate, shares)
+	returnCoins := sdk.Coins{{k.GetParams(ctx).BondDenom, returnAmount}}
+	k.coinKeeper.AddCoins(ctx, bond.DelegatorAddr, returnCoins)
+
+	// lastly if an revoke candidate if necessary
+	if revokeCandidacy {
+
+		// change the share types to unbonded if they were not already
+		if candidate.Status == Bonded {
+			k.bondedToUnbondedPool(ctx, candidate)
+		}
+
+		// lastly update the status
+		candidate.Status = Revoked
+	}
+
+	// deduct shares from the candidate
+	if candidate.Liabilities.IsZero() {
```

Comment:
This is not the check in the spec; this checks bonded delegate shares (`candidate.Liabilities`), but the spec checks global stake shares (`candidate.Assets`).